### PR TITLE
Fix `IndexDefect` errors in httpclient on invalid/weird headers

### LINKED ISSUE #19261: `IndexDefect` in httpclient because of a header spanning multiple lines
I was doing some scanning stuff and found a host that replies with one header that spans multiple lines. Chromium seems to be happy with it, but `httpclient` does an out of bounds access which results in IndexDefect.

### Example
No usable-by-default Nim repro code yet, but here's the whole response from the server:
```html
> GET / HTTP/1.1
> Host: [REDACTED]
> User-Agent: curl/7.80.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< Server: nginx
< Date: Fri, 17 Dec 2021 00:23:58 GMT
< Content-Type: text/html
< Content-Length: 178
< Connection: keep-alive
< Keep-Alive: timeout=5
< Location: ./login.html
< Content-Security-Policy: default-src 'self';
<                                     script-src 'self'
<                                                'unsafe-inline'
<                                                'unsafe-eval';
<                                     img-src 'self'
<                                             data:;
<                                     style-src 'self'
<                                               'unsafe-inline';
<                                     font-src 'self';
<                                     child-src 'self';
<                                     object-src 'none'
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-XSS-Protection: 1; mode=block
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< 
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

Notice how `Content-Security-Policy` spans multiple lines.

### Current Output
```
/home/dian/Projects/PrivWeb/scratchpad.nim(47) scratchpad
/home/dian/Things/nim/lib/pure/httpclient.nim(1188) getContent
/home/dian/Things/nim/lib/pure/httpclient.nim(1183) get
/home/dian/Things/nim/lib/pure/httpclient.nim(1110) request
/home/dian/Things/nim/lib/pure/httpclient.nim(1057) requestAux
/home/dian/Things/nim/lib/pure/httpclient.nim(846) parseResponse
/home/dian/Things/nim/lib/system/fatal.nim(53) sysFatal
Error: unhandled exception: index 53 not in 0 .. 52 [IndexDefect]
```

### Expected Output
No error from `httpclient`, and it should parse that header into a single line, like Chromium:
image

### Additional Information
```
$ nim -v
Nim Compiler Version 1.7.1 [Linux: amd64]
Compiled at 2021-12-17
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 0d0c249074d6a1041de16108dc247396efef5513
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ghost**
Continuation of https://github.com/nim-lang/Nim/pull/19262

Fixes https://github.com/nim-lang/Nim/issues/19261

The parsing code is still too lenient (e.g. it will happily parse header names with spaces in them, which is outright invalid by the spec), but I didn't want to touch it beyond the simple changes to make sure that `std/httpclient` won't throw `IndexDefect`s like it does now on those cases:
- Multiline header values
- No colon after the header name
- No value after the header name + colon

One question remains - should I keep `toCaseInsensitive` exported in `httpcore` or just copy-paste the implementation?

### CODE REVIEW

**Comment by ghost** [comment]
A simple test I made specifically for multiline headers:
```nim
import std/[asyncdispatch, asyncnet, os, strutils, strformat]

const csp = """default-src 'self';
                                     script-src 'self'
                                                'unsafe-inline'
                                                'unsafe-eval';
                                     img-src 'self'
                                             data:;
                                     style-src 'self'
                                               'unsafe-inline';
                                     font-src 'self';
                                     child-src 'self';
                                     object-src 'none'"""

proc handleClient(client: AsyncSocket) {.async.} =
  let line = await client.recvLine()
  var toAdd = ""
  if "case1" in line:
    # multiline header
    toAdd = &"Content-Security-Policy: {csp}\r\n"
  elif "case2" in line:
    # no colon after the header name!
    toAdd = "X-XSS-Protection 1; mode=block\r\n"
  elif "case3" in line:
    # no value after colon
    toAdd = "X-Xss-Protection:\r\n"
  let
    body = "<h1>Hello World</h1>"
    headers = "HTTP/1.1 200 OK\r\n" &
               "Server: nginx\r\n" &
               # this should work fine
               "Some-Empty-Header: \r\n"&
               "Content-Type: text/html\r\n" &
               "Content-Length: " & $body.len & "\r\n" &
               toAdd &
               "\r\n"
  await client.send(headers & body)
  client.close()

proc main {.async.} =
  let server = newAsyncSocket()
  server.bindAddr(Port(8080))
  server.listen()
  defer: server.close()
  echo "Running on http://localhost:8080/"
  while true:
    let client = await server.accept()
    asyncCheck handleClient(client)

waitFor main()
```

With client:
```nim
import std/[httpclient, strformat, httpcore, strutils, strformat, unittest]

proc testRequest(url: string): Response = 
  echo fmt"Requesting {url}"
  let c = newHttpClient("curl/8.4.0")
  let data = c.get(url)
  return data

const
  server = "http://localhost:8080"

const cspExpected = """default-src 'self';
                                     script-src 'self'
                                                'unsafe-inline'
                                                'unsafe-eval';
                                     img-src 'self'
                                             data:;
                                     style-src 'self'
                                               'unsafe-inline';
                                     font-src 'self';
                                     child-src 'self';
                                     object-src 'none'"""

suite "Test Requests":
  test "Basic empty header - OK":
    let response = testRequest(fmt"{server}/")
    check response.status == "200 OK"
    check response.headers["Some-Empty-Header"] == ""
  
  test "Multiline header - OK":
    let response = testRequest(fmt"{server}/case1")
    check response.status == "200 OK"
    let cspHeader = response.headers.getOrDefault("Content-Security-Policy")
    check cspExpected == cspHeader

  test "No colon after - should fail":
    try:
      let response = testRequest(fmt"{server}/case2")
    except ProtocolError as e:
      check "no colon after" in e.msg 

  test "No value after header + colon - should fail":
    try:
      let response = testRequest(fmt"{server}/case3")
    except ProtocolError as e:
      check "no header value after" in e.msg
```

File: lib/pure/httpclient.nim [hunk 1]
```diff
@@ -890,16 +891,32 @@ proc parseResponse(client: HttpClient | AsyncHttpClient,
...
+      if line[0] in {' ', '\t'}:
+        # Check if it's a multiline header value, if so, append to the header we're currently parsing
+        # This works because a line with a header must start with the header name without any leading space
+        # See https://datatracker.ietf.org/doc/html/rfc7230, section 3.2 and 3.2.4
+        # Multiline headers are deprecated in the spec, but it's better to parse them than crash
+        if lastHeaderName == "":
+          # This should ideally throw an error, but old httpclient actually treated lines like this
+          # as a header with an empty name and didn't crash
+          #httpError("Invalid headers - received multiline header value without previous header")
+          discard
```
Comment (ghost) [inline lib/pure/httpclient.nim]: Should I change this or not? I just fear that some extremely weird servers might actually send some unparsed lines in the response, and httpclient before just ignored those (added to the empty header)

Comment (ghost) [reply lib/pure/httpclient.nim]: FWIW, Python's requests also ignores this, but curl throws an error with "curl: (8) Header without colon"

File: lib/pure/httpclient.nim [hunk 2]
```diff
@@ -911,7 +911,9 @@ proc parseResponse(client: HttpClient | AsyncHttpClient,
         if line.len == le: httpError("Invalid headers - no colon after header name")
         inc(linei, le) # Skip the parsed header name
         inc(linei) # Skip :
-        if linei == line.len: httpError("Invalid headers - no header value after colon")
+        # Here linei is smaller or equal to line.len, so the slice below should work fine
+        # If we want to enforce the HTTP spec later, do this:
+        #if linei == line.len: httpError("Invalid headers - no header value after colon")
```
Comment (ghost) [inline lib/pure/httpclient.nim]: It's not spec compliant, but I still commented this out for discussion - at least both Curl and Python seem to happily take headers that have a CRLF right after the header name + colon without any value and then show it as a header with an empty value.

File: lib/pure/httpclient.nim (hunk 1 above)
Comment (Araq) [reply lib/pure/httpclient.nim]: Follow Python's requests then.

File: lib/pure/httpclient.nim (hunk 2 above)
Comment (Araq) [reply lib/pure/httpclient.nim]: It's fine to copy Curl's and Python's behavior.

Comment (Varriount) [reply lib/pure/httpclient.nim]: Makes me wonder if we should set up fuzzing tests for certain modules (to be run manually, at least).

**Comment by ghost** [comment]
I updated the comments so that we just don't crash but not really enforce the spec in any way. Last question - should I keep `toCaseInsensitive` exported (it's private right now) to use it, or copy the implementation?

File: lib/pure/httpcore.nim [hunk 3]
```diff
@@ -126,7 +126,7 @@ func toTitleCase(s: string): string =
     result[i] = if upper: toUpperAscii(s[i]) else: toLowerAscii(s[i])
     upper = s[i] == '-'
 
-func toCaseInsensitive(headers: HttpHeaders, s: string): string {.inline.} =
+func toCaseInsensitive*(headers: HttpHeaders, s: string): string {.inline.} =
```
Comment (Araq) [inline lib/pure/httpcore.nim:129]:
```suggestion
func toCaseInsensitive*(headers: HttpHeaders, s: string): string {.inline.} =
  ## For internal usage only. Do not use.
```

**Pull request merged**